kernel/src/process.rs

```diff
@@ -982,31 +1043,105 @@ impl<C: Chip> ProcessType for Process<'_, C> {
     }
 
     fn set_fault_state(&self) {
-        self.state.update(State::Fault);
-
         match self.fault_response {
             FaultResponse::Panic => {
                 // process faulted. Panic and print status
+                self.state.update(State::Faulted);
                 panic!("Process {} had a fault", self.process_name);
             }
-            FaultResponse::Restart(_) => {
-                self.restart(State::StoppedFaulted);
+            FaultResponse::Restart(restart_policy) => {
+                // Apply the process policy for whether to restart
+                // on a fault. We sometimes don't want to (e.g., too
+                // many faults).
+                if restart_policy.should_restart(self) {
+                    self.try_restart(COMPLETION_FAULT);
```

Comment:
We need to clarify how restart policies are intended to be handled in the kernel, and on a per-process basis. I know I've brought this up before, but reading the comments and code and I'm not clear what the intention is, or how we would explain that to board authors if they have to decide what policies to select.

When a process is put in the "fault" state via this fn (`set_fault_state`), we check the per-process fault policy selected by the board, and if the policy is restart we check here if the kernel should even attempt to restart the process. `should_restart` can inspect process state to make that decision. What's confusing is that the description of `try_restart` indicates it too should implement a policy, and cites examples which are the same as current `FaultResponse::Restart(restart_policy)` policies. So there is a policy on fault handling, and then a potential new policy on restarts, both of which cover restarts. The difference is only the former policy would apply to processes that call exit-restart.

I don't think we need to write any code, but we do need to clarify what the intended design is, and we should do it in this PR because this is where we are introducing the new ambiguity that does not currently exist.